dags/top-downloads-from-sowflake.py

```diff
@@ -83,25 +93,66 @@
+    """Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        name: The name of the project
+        project: The ID of the project
+        downloads_per_project: The number of downloads per project
+        number_of_unique_users_downloaded: The number of unique users who downloaded
+        data_download_size: The size of the data downloaded
+
+    """
+
+    name: str
+    project: str
+    downloads_per_project: int
+    number_of_unique_users_downloaded: int
+    data_download_size: float
+
+
 @dag(**dag_config)
 def snowflake_top_downloads_to_slack() -> None:
     """Execute a query on Snowflake and report the results to a slack channel."""
 
     @task
-    def get_top_downloads_from_snowflake(**context) -> str:
-        """Execute the query on Snowflake and return the results as a string."""
+    def get_top_downloads_from_snowflake(**context) -> List[DownloadMetric]:
+        """Execute the query on Snowflake and return the results."""
         snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
         ctx = snow_hook.get_conn()
         cs = ctx.cursor()
         cs.execute(QUERY)
         top_downloaded_df = cs.fetch_pandas_all()
 
-        message = ":synapse: Top Downloaded Public Synapse Projects Yesterday (UTC)!\n\n"
-        for index, row in top_downloaded_df.iterrows():
-            name = row["NAME"]
-            url = row["PROJECT"]
-            downloads = row["DOWNLOADS_PER_PROJECT"]
-            message += f"{index+1}. <{url}|{name}> - {downloads} downloads\n\n"
+        metrics = []
+        for _, row in top_downloaded_df.iterrows():
+            metrics.append(
+                DownloadMetric(
+                    name=row["NAME"],
+                    project=row["PROJECT"],
+                    downloads_per_project=row["DOWNLOADS_PER_PROJECT"],
+                    number_of_unique_users_downloaded=row[
+                        "NUMBER_OF_UNIQUE_USERS_DOWNLOADED"
+                    ],
+                    data_download_size=row["DATA_DOWNLOAD_SIZE"],
+                )
+            )
+        return metrics
+
+    @task
+    def generate_top_downloads_message(metrics: List[DownloadMetric]) -> str:
+        """Generate the message to be posted to the slack channel."""
+        message = ":synapse: Top Downloaded Public Synapse Projects Yesterday!\n\n"
+        for index, row in enumerate(metrics):
+            if row.data_download_size:
+                size_string = (
+                    f"{row.data_download_size:.{SIZE_ROUNDING}f} {BYTE_STRING}"
+                )
+            else:
+                size_string = f"< {0:.{SIZE_ROUNDING}f}5 {BYTE_STRING}"
+            message += f"{index+1}. <https://www.synapse.org/#!Synapse:{row.project}|{row.name}> - {row.downloads_per_project} downloads, {row.number_of_unique_users_downloaded} unique users, {size_string} {BYTE_STRING} egressed\n\n"
         message += "One download is a user downloading a file once\n"
```

Comment:
Can this be edited? it's not just files that are being tracked.

Reply:
Updated with information in ticket AC: https://github.com/Sage-Bionetworks-Workflows/orca-recipes/pull/37/commits/9e88498696e1e20b5d767441ae9f78afcaa9ac5f